src/UpshotAdapter.sol

```diff
@@ -114,7 +95,41 @@ contract UpshotAdapter is IUpshotAdapter, Ownable2Step, EIP712 {
     ///@inheritdoc IUpshotAdapter
     function verifyData(
         UpshotAdapterNumericData calldata nd
-    ) external payable override returns (uint256 numericValue) {
+    ) external override returns (uint256 numericValue) {
+        uint256 topicId;
+        address[] memory dataProviders;
+        bytes calldata extraData;
+
+        (numericValue, topicId, dataProviders, extraData) = _verifyData(nd);
+
+        topicValue[topicId][extraData] = TopicValue({
+            recentValue: _toUint192(numericValue),
+            recentValueTime: _toUint64(block.timestamp)
+        });
+
+        emit UpshotAdapterV2AdapterVerifiedData(topicId, numericValue, dataProviders, extraData);
+    }
+
+    ///@inheritdoc IUpshotAdapter
+    function verifyDataViewOnly(
+        UpshotAdapterNumericData calldata nd
+    ) external view override returns (uint256 numericValue) {
+        (numericValue,,,) = _verifyData(nd);
+    }
+
+    /**
+     * @notice Verify the data provided by the data providers
+     * 
+     * @param nd The data to verify
+     */
+    function _verifyData(
+        UpshotAdapterNumericData calldata nd
```

Comment:
For `verifyData()`, `verifyDataViewOnly()`, and `_verifyData()`, the input should be memory. The reason is because the entry function the user interface with, say `quote()`, will accept a generic parameter with a `bytes` type, say `bytes loanContext`. It contains many things, one of which might be numeric data. Hence, we need to do something like `abi.decode(loanContext, (UpshotAdapterNumericData))` to get `nd` before calling `verifyData()` with it. At this point, `nd` cannot be `calldata`. Without changing `nd` to `memory`, we can't use it.